src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SafeHandle.cs

```diff
@@ -0,0 +1,258 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.ConstrainedExecution;
+using System.Threading;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    // This implementation does not employ critical execution regions and thus cannot
+    // reliably guarantee handle release in the face of thread aborts.
+
+    /// <summary>Represents a wrapper class for operating system handles.</summary>
+    public abstract partial class SafeHandle : CriticalFinalizerObject, IDisposable
+    {
+        // IMPORTANT:
+        // - Do not add or rearrange fields as the EE depends on this layout,
+        //   as well as on the values of the StateBits flags.
+        // - The EE may also perform the same operations using equivalent native
+        //   code, so this managed code must not assume it is the only code
+        //   manipulating _state.
+
+        /// <summary>Specifies the handle to be wrapped.</summary>
+        [SuppressMessage("Microsoft.Security", "CA2111:PointersShouldNotBeVisible")]
+        protected IntPtr handle;
+        /// <summary>Combined ref count and closed/disposed flags (so we can atomically modify them).</summary>
+        private volatile int _state = StateBits.RefCountOne; // Ref count 1 and not closed or disposed.
+        /// <summary>Whether we can release this handle.</summary>
+        private readonly bool _ownsHandle;
+        /// <summary>Whether constructor completed.</summary>
+        private volatile bool _fullyInitialized;
+
+        /// <summary>Bitmasks for the <see cref="_state"/> field.</summary>
+        /// <remarks>
+        /// The state field ends up looking like this:
+        ///
+        ///  31                                                        2  1   0
+        /// +-----------------------------------------------------------+---+---+
+        /// |                           Ref count                       | D | C |
+        /// +-----------------------------------------------------------+---+---+
+        /// 
+        /// Where D = 1 means a Dispose has been performed and C = 1 means the
+        /// underlying handle has been (or will be shortly) released.
+        /// </remarks>
+        private static class StateBits
+        {
+            public const int Closed = 0b01;
+            public const int Disposed = 0b10;
+            public const int RefCount = unchecked(~0b11); // 2 bits reserved for closed/disposed; ref count gets 30 bits
+            public const int RefCountOne = 1 << 2;
+        };
+
+        /// <summary>Creates a SafeHandle class.</summary>
+        protected SafeHandle(IntPtr invalidHandleValue, bool ownsHandle)
+        {
```

Comment:
@jkotas, do you think we need to do anything with RuntimeHelpers.PrepareMethod for the virtual calls that'll get made?

Reply:
I do not think we need to do anything.

Reply:
Great.